### PR TITLE
Initial front-end implementation of submarine

### PR DESCRIPTION
**Pull request by ngorog**

### CODE REVIEW

**Comment by ktran0918** [comment]
Submarine is now submersible. Here are the implementation notes:
- The submerse state of the sub is now represented by the case of the ship kind. "SUBMARINE" represents a submersed sub, "submarine" otherwise. This applies to both the string parsing from the front-end and the back-end Submarine class implementation. (Note that there's no boolean to indicate submersed/surfaced anymore).
- The UI allows the user to press "u" key to toggle between submersed/surfaced placement. The former state is shown as a green outline, during placement for now. When it's placed it'll appear the same, albeit not implemented currently.
- BUG: I cannot get the submersed sub to pass the collision check. What I did was OR-ing the validLocation () with ship.shipKing == "SUBMARINE"; in other words, it's a valid ship placement if there's no collision or the ship type is a submersed sub. The function still returns false. If anyone could help me out with looking into this, I'd appreciate it.
- One more thing to implemented: a prompt in the UI to inform the user that the shortcut "u" is used to toggle the submersed state. @dongrogers You want to help me out with this?

**Review by barnesjt — APPROVED** [review]
Here's what I've come up with:
### Back End
- Placing a surfaced submarine causes all attacks to be rejected by the back-end.
- The odd submarine shape is not accounted for when detecting collisions with the board-edge or other ships during placement.
- A submerged submarine allows the attack phase to proceed, but cannot be placed under other ships.

### game.js (Front-end)
- The hover listener on the front-end does not properly display a submarine outline.
- The submarine outline does not rotate with r (instead changes to a single highlighted square).
- When rotated for placement, a submarine is not actually rotated.


While I believe there is more work that could be done on this particular issue, I believe it is advantageous to approve this pull request so that we can step forward with development.

Also, if I commit code to this branch, then Brad would be the only person that could then review this PR.
Lets get this in and work towards fixes.

**Comment by barnesjt** [comment]
Quick note that this was for #50 Assigned to Khoa AND #53 Assigned to Rogers.

File: src/main/java/assets/css/custom.css [hunk 1]
```diff
@@ -92,6 +92,12 @@ td:hover {
 	background-color: black;
 }
 
+.submerged {
+    background-color: transparent !important;
```
Comment (barnesjt) [inline src/main/java/assets/css/custom.css:96]: This doesn't work,

File: src/main/java/assets/game.js [hunk 2]
```diff
@@ -271,7 +278,12 @@ function place(size) {
                 // ship is over the edge; let the back end deal with it
                 break;
             }
-            cell.classList.toggle("placed");
+            if (isSubmerged) {
+                cell.classList.remove("placed");
```
Comment (barnesjt) [inline src/main/java/assets/game.js:282]: Removing the placed class directly was kinda messing up the toggling when switching around

File: src/main/java/assets/game.js [hunk 3]
```diff
@@ -365,6 +377,27 @@ function doShipPlacement() {
...
         prompt.appendChild(rotateKey);
+    } else if (placedShips==3){
+        shipType = "submarine";
+        registerCellListener(place(4));
+        document.addEventListener('keypress', function(e) {
+            if (e.which == 85 || e.which == 117) {
+                isSubmerged = !isSubmerged;
+                redrawGrid();
+
+                // "SUBMARINE" indicates a submersed sub; "submarine" means a surfaced sub
```
Comment (barnesjt) [inline src/main/java/assets/game.js:388]: This was good to put in the event listener, I was originally thinking elsewhere, but this works out very well actually.

File: src/main/java/assets/game.js [hunk 4]
```diff
@@ -385,8 +418,10 @@ function initGame() {
           if (shipType=="MINESWEEPER" && isSetup) registerCellListener(place(2));
           else if (shipType=="DESTROYER" && isSetup) registerCellListener(place(3));
           else if (shipType=="BATTLESHIP" && isSetup) registerCellListener(place(4));
+          else if (shipType=="SUBMARINE" && isSetup) registerCellListener(place(4));
```
Comment (barnesjt) [inline src/main/java/assets/game.js:421]: Using the case as a flag wasn't mixing well with this conditional

File: src/main/java/controllers/ApplicationController.java [hunk 5]
```diff
@@ -21,7 +22,15 @@ public Result newGame() {
 
     public Result placeShip(Context context, PlacementGameAction g) {
         Game game = g.getGame();
-        Ship ship = new Ship(g.getShipType());
+        Ship ship;
+        // submerged flag is determined by case -- see doShipPlacement() in game.js
+        if(g.getShipType().equals("SUBMARINE") || g.getShipType().equals("submarine")) {
```
Comment (barnesjt) [inline src/main/java/controllers/ApplicationController.java:27]: Actually a .toUpperCase() on the string works well for this.

File: src/main/java/cs361/battleships/models/Board.java [hunk 6]
```diff
@@ -27,7 +27,7 @@ public boolean placeShip(Ship ship, int x, char y, boolean isVertical) {
 		if(this.getShips().size()>0) {
 			if (!validShipType(ship)) return false;
 		}
-		if (validLocation(shipSize, x, y, isVertical)) {
+		if (validLocation(shipSize, x, y, isVertical) || ship.getKind() == "SUBMARINE") {
```
Comment (barnesjt) [inline src/main/java/cs361/battleships/models/Board.java:30]: This skips validation checking for the submarine, definitely changing this.

File: src/main/java/cs361/battleships/models/Game.java [hunk 7]
```diff
@@ -13,9 +13,20 @@
 	DO NOT change the signature of this method. It is used by the grading scripts.
 	 */
     public boolean placeShip(Ship ship, int x, char y, boolean isVertical) {
-        boolean successful = playersBoard.placeShip(new Ship(ship.getKind()), x, y, isVertical);
+        boolean successful;
+        String shipKind = ship.getKind();
+        if(shipKind.equals("SUBMARINE") || shipKind.equals("submarine")) {
```
Comment (barnesjt) [inline src/main/java/cs361/battleships/models/Game.java:18]: Again, we can avoid the OR here

File: src/main/java/cs361/battleships/models/Submarine.java [hunk 8]
```diff
@@ -7,10 +7,10 @@
 
 public class Submarine extends Ship {
 
-	public Submarine() {
+	public Submarine(String kind) {
```
Comment (barnesjt) [inline src/main/java/cs361/battleships/models/Submarine.java:10]: This is probably a good place to restore the proper name and begin storing the submerged option as it's own variable so it's easier to access by other functions

**Pull request merged**